docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+- [Constants](#module-configurations)
+- [Transfer](#transfer-object)
+
+# Layers
+
+An application layer can have up to four logical layers with clear purpose and communication rules.
+- **Presentation Layer**: contains twig-templates, JS and CSS files
+- **Communication Layer**: contains controllers, forms, tables and plugins
+- **Business Layer**: contains the business logic of a module
+- **Persistence Layer**: contains queries, entities and the schema of a module
+
+## Presentation Layer responsibilities
+- Responsible for handling the user interface (UI) presentation.
+- Contains frontend elements, such as HTML, Twig templates, JavaScript (JS), TypeScript, Cascading Style Sheets (CSS) files, and so on.
+- Handles user interactions and input validations on the client side.
+- Interacts with the [Communication Layer](#communication-layer-responsibilities) to retrieve necessary data for display.
+
+## Communication Layer responsibilities
+- Acts as an intermediary between the [Presentation Layer](#presentation-layer-responsibilities) and the [Business Layer](#business-layer-responsibilities).
+- Contains [controllers](#controller) responsible for handling HTTP requests and responses.
+- Contains [plugins](#plugin) responsible for flexible, overarching requests and responses.
+- Manages form processing and validation.
+- Handles [routing and dispatching requests](#provider--router) to appropriate [controllers](#controller).
+- Interacts with the [Business Layer](#business-layer-responsibilities) to perform business operations.
+
+## Business Layer responsibilities
+- Contains the main business logic.
+- Implements business rules and processes.
+- Performs data manipulation, calculations, and validation.
+- Interacts with the [Persistence Layer](#persistence-layer-responsibilities) to read and write data.
+- Independent of the user interface and communication protocols.
+
+## Persistence Layer responsibilities
+- Responsible for data storage and retrieval.
+- Contains queries (via [Entity Manager](#entity-manager) or [Repository](#repository)), [entities](#entity) (data models), and [database schema definitions](#persistence-schema).
+- Handles database operations such as CRUD (`Create`, `Read`, `Update`, `Delete`).
+- Ensures data integrity and security.
+
+# Components
+
+**Conventions**
+- The components MUST be placed according to the corresponding [application layer’s](#application-layers) directory architecture  in order to come in effect.
+- The components MUST inherit from the [application layer](#application-layers) corresponding abstract class in `Kernel` module.
+- The components MUST be stateless to be deterministic and easy to comprehend.
+
+<details><summary markdown='span'>Additional Conventions for Module Development</summary>
+- The components MUST be extended directly from the [application layer](#application-layers) corresponding abstract class in `Kernel` module.<br/>
+</details>
+
+## Client facade
```

Comment:
Can we just merge all facade use cases into one description like controllers and dependency providers?